### DISCUSSION TITLE
Is integers until 256 a Singleton like in Python ?

### DISCUSSION THREAD

**Discussion by mfranzon** [modular/modular#155] [Q&A] [answered]
In python, number until 256 are considered a Singleton, for example:

```python
a = 256
b = 256
id(a) == id(b) # return True 
c = 257
d = 257
id(c) == id(d) # return False
```
1. Is in Mojo the same behaviour ?
2. Is there an equivalent function like python id in mojo?

Thank you!

**Comment by czheo** [comment, marked as answer]
Mojo is a compiled language so I think it will be very different. Here's my understanding.

Now Mojo only has `Int` which is basically a signed 64 bit integer, while Python `int` is an object that support arbitrarily large size. So primary type vs object are very different at the first place.
Imagine you define 2 mutable variables in Mojo like below.
```
var a : Int = 256
var b : Int = 256
```
They may have different memory addresses, since you will mutate the int values stored in those addresses independently when you change them. If you use `let` to make them immutable, I guess Mojo may eliminate one of them/put them into the same address in the read-only data segment during compilation/optimization. (Even with `var` the code might be optimized, if the compiler figures out you didn't actually mutate them.)

Since Python equivalent `int` is not implemented yet, we don't know how it will look like.

**Reply by mfranzon** [reply]
I am agree, also in general if you want to achieve better memory usage, singleton implementation is not the best choice!

**Reply by lattner** [reply]
nice answer!
